Interrogator.xUnit/NotConcurrentAttribute.cs

```diff
@@ -0,0 +1,139 @@
+
+		public enum ConcurrencyScope
+		{
+			Class,
+			ClassHierarchy,
+			Namespace,
+			Assembly
+		}
+
+		public string GroupName { get; }
+		public ConcurrencyScope Scope { get; }
+
+		public NotConcurrentAttribute() : this("", ConcurrencyScope.Class) { }
+		public NotConcurrentAttribute(string groupName) : this(groupName, ConcurrencyScope.Class) { }
+
+		public NotConcurrentAttribute(ConcurrencyScope scope) : this("", scope) { }
+
+		public NotConcurrentAttribute(string groupName, ConcurrencyScope scope) : base(null)
+		{
+			GroupName = $"{groupName}_{scope}"; // Adding scope to group name ensures that different scoped attributes don't collide
+			Scope = scope;
+		}
+
+		internal override Result<Option<MethodInfo>, string> TryGetMethod(Type containingType, MemberInfo member, MethodInfo[] testMethods)
+		{
+			var result = (Type ?? containingType)
+				.TryGetMethod(member.Name, ParameterTypes)
+				.Select(currentMethod => GetPreviousInChain(currentMethod, containingType, Scope, GroupName, testMethods));
+			return result;
+		}
+
+		private static Option<MethodInfo> GetPreviousInChain(MethodInfo currentMethod, Type type, ConcurrencyScope scope, string groupName, MethodInfo[] executingMethods)
+		{
+			var methodSet = new HashSet<MemberInfo>(executingMethods);
+			var methodsToExecute = GetSameGroupMethods(scope, type, groupName).Where(methodSet.Contains).ToArray();
+
+			return GetPreviousMethod(currentMethod, methodsToExecute);
+		}
+
+		private static MethodInfo[] GetSameGroupMethods(ConcurrencyScope scope, Type type, string groupName)
+		{
+			if (!_methodInfoDictionary.ContainsKey(type))
+			{
+				IEnumerable<Type> types;
+				switch (scope)
+				{
+					case ConcurrencyScope.Class:
+						types = new[] { type };
+						break;
+					case ConcurrencyScope.Assembly:
+						types = GetTypesInAssembly(type);
+						break;
+					case ConcurrencyScope.Namespace:
+						types = GetTypesInNamespace(type);
+						break;
+					case ConcurrencyScope.ClassHierarchy:
+						types = GetTypesInClassHierarchy(type);
+						break;
+					default:
+						types = Array.Empty<Type>();
+						break;
+				}
+				_methodInfoDictionary[type] = GetMethodsInTypes(groupName, types.ToArray());
+			}
+
+			return _methodInfoDictionary[type];
+		}
+
+		private static MethodInfo[] GetMethodsInTypes(string groupName, params Type[] types)
+			=> types
+				.OrderTypes()
+				.SelectMany(t => GetGroupMethodsInType(t, groupName))
+				.ToArray();
+
+		private static IEnumerable<MethodInfo> GetGroupMethodsInType(Type type, string groupName)
+			=> type
+				.GetMethods()
+				.Where(m => m.GetCustomAttributes(true).OfType<NotConcurrentAttribute>().Any(x => x.GroupName == groupName))
+				.OrderBy(GetMethodKey);
+
+		private static string GetMethodKey(MethodInfo method)
+			=> $"{method.Name}-{String.Join("-", method.GetParameters().Select(p => p.ParameterType.FullName))}";
+
+		private static IEnumerable<Type> GetTypesInClassHierarchy(Type type)
+		{
```

Comment:
If you have class A that contains classes B and C, and then you call this on class C, you won't actually get class B back will you?

Reply:
Hmm, good point. Though, I think in your example we would only want A and C to be considered.  I think the issue is more if you try to run on B and C at the same time.